x-pack/plugins/infra/public/containers/logs/log_position/log_position_state.ts

```diff
@@ -79,26 +114,94 @@ export const useLogPositionState: () => LogPositionStateParams & LogPositionCall
     [startKey, endKey]
   );
 
+  // Allow setting `startDate` and `endDate` separately, or together
+  const updateDateRange = useCallback(
+    (newDateRange: Partial<DateRange>) => {
+      // Prevent unnecessary re-renders
+      if (!('startDateExpression' in newDateRange) && !('endDateExpression' in newDateRange)) {
+        return;
+      }
+
+      const nextStartDateExpression =
+        newDateRange.startDateExpression || dateRange.startDateExpression;
+      const nextEndDateExpression = newDateRange.endDateExpression || dateRange.endDateExpression;
+
+      if (!isValidDatemath(nextStartDateExpression) || !isValidDatemath(nextEndDateExpression)) {
+        return;
+      }
+
+      // Dates are valid, so the function cannot return `null`
+      const nextStartTimestamp = datemathToEpochMillis(nextStartDateExpression)!;
+      const nextEndTimestamp = datemathToEpochMillis(nextEndDateExpression, 'up')!;
+
+      // Reset the target position if it doesn't fall within the new range.
+      if (
+        targetPosition &&
+        (nextStartTimestamp > targetPosition.time || nextEndTimestamp < targetPosition.time)
+      ) {
+        jumpToTargetPosition(null);
+      }
+
+      setDateRange(previousDateRange => {
+        return {
+          startDateExpression:
+            newDateRange.startDateExpression || previousDateRange.startDateExpression,
+          endDateExpression: newDateRange.endDateExpression || previousDateRange.endDateExpression,
+        };
+      });
+      updateTimestamps();
+    },
+    [dateRange, targetPosition, updateTimestamps]
+  );
+
+  // `lastUpdate` needs to be a dependency for the timestamps.
+  // ESLint complains it's unnecessary, but we know better.
+  /* eslint-disable react-hooks/exhaustive-deps */
+  const startTimestamp = useMemo(() => datemathToEpochMillis(dateRange.startDateExpression), [
+    dateRange.startDateExpression,
+    timestampsLastUpdate,
+  ]);
+
+  // endTimestamp needs to be synced to `now` to allow auto-streaming
+  const endTimestampDep =
+    dateRange.endDateExpression === 'now' ? Date.now() : dateRange.endDateExpression;
+  const endTimestamp = useMemo(() => datemathToEpochMillis(dateRange.endDateExpression, 'up'), [
+    endTimestampDep,
+    timestampsLastUpdate,
+  ]);
+  /* eslint-enable react-hooks/exhaustive-deps */
```

Comment:
Do we want to prefer the single-line override here so we don't accidentally disable it for more than the intended lines when changes are introduced later?

```suggestion
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const endTimestamp = useMemo(() => datemathToEpochMillis(dateRange.endDateExpression, 'up'), [
    endTimestampDep,
    timestampsLastUpdate,
  ]);
```